Add page.on('metric') docs (#1807)
grafana/k6-docs: docs/sources/k6/next/javascript-api/k6-browser/metricmessage.md

Reviewer: I want to point out that the other files in the documentation use hyphens to separate words like `metric-message` instead of `metricmessage`. One existing example is `console-message`. Do you think we should rename this file for consistency as well?

Author: I did see that, but i believe that is the outlier. All other doc file names do not contain a hyphen. I'll stick with the one without.

Reviewer 2: It seems like there's some inconsistency in the naming across the JS API section, but I agree with @ankur22 here that it's okay to use it without the hyphen. 👍

---
title: 'MetricMessage'
slug: 'metricmessage'
description: 'Browser module: MetricMessage Class'
weight: 03
---

# MetricMessage

> `MetricMessage` objects are dispatched by page via registering a handler with [page.on('metric')](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-browser/page/on). For each metric that it measured and emitted for the page, k6 browser delivers it to the registered handlers.

Reviewer: For clarity, I would first explain what a `MetricMessage` is and then where it's emitted and used. Something like:

```suggestion
A `MetricMessage` object allows tagging `page` metrics.

`MetricMessage` objects are dispatched by the page when registered a handler with [page.on('metric')](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-browser/page/on). For each metric that is measured and emitted for the page, the k6 browser delivers a `MetricMessage` object to the registered handlers, allowing them to pattern match and tag metrics.
```

Author: Thanks, I've used your suggestion and altered it a bit to:

```suggestion
A `MetricMessage` object allows tagging of metrics that are measured and emitted for the page.

`MetricMessage` objects are dispatched by the page when a handler is registered with [page.on('metric')](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-browser/page/on). For each metric that is measured and emitted for the page, the k6 browser delivers a `MetricMessage` object to the registered handlers, allowing them to pattern match and tag metrics.
```

## tag

> The Metric Message exposes a single method, `tag`. It will match the given `matches` with the current metric's url and name tags. When a match is found it will use `name` to replace the existing url and name tag values.

Reviewer: I'd remove the "single method" part as it's already clear from the context.

```suggestion
The `Metric.tag` method matches the given `matches` with the current metric's url and name tags. When a match is found, it will use `name` to replace the existing url and name tag values.
```

> Doing this will help group metrics with disparate url and name tags, which are in fact referencing the same resource, so that a correlation can be found over time and preventing high cardinality issues in the backend database that is storing the metrics.

Reviewer: I would remove the backend part from this text as it's related to k6 internals.

Author: Sure, updated in 98464b654fd8c72ea439cd061636d4d7489f0a1f

---
title: 'MetricMessage'

> slug: 'metricmessage'

Reviewer 2: ```suggestion
```

If the URL is going to be the same name as the filename, you don't need the `slug` property here. :)

Author: Sure, removed in 3baf38691cea0e84ba92bad652d60e47bed7117b

description: 'Browser module: MetricMessage Object'
weight: 03
---

# MetricMessage

A `MetricMessage` object allows tagging of metrics that are measured and emitted for the page.

> `MetricMessage` objects are dispatched by the page when a handler is registered with [page.on('metric')](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-browser/page/on). For each metric that is measured and emitted for the page, the k6 browser delivers a `MetricMessage` object to the registered handlers, allowing them to pattern match and tag metrics.

Reviewer 2: ```suggestion
`MetricMessage` objects are dispatched by the page when a handler is registered with [page.on('metric')](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-browser/page/on). For each metric that is measured and emitted for the page, the k6 browser module delivers a `MetricMessage` object to the registered handlers, allowing them to pattern match and tag metrics.
```

## tag

> The `tag` method matches the given `matches` with the current metric's url and name tags. When a match is found, it will use `name` to replace the existing url and name tag values.

Reviewer 2: ```suggestion
The `tag` method matches the given `matches` with the current metric's url and name tags. When a match is found, it will use `name` to replace the existing URL and name tag values.
```

> Doing this helps to group metrics with disparate url and name tags, which are, in fact, referencing the same resource so that a correlation can be found over time and to reduce the cardinality of the metrics.

Reviewer 2: ```suggestion
Doing this helps group metrics with different URL and name tags, which are, in fact, referencing the same resource so that a correlation can be found over time and to reduce the cardinality of the metrics.
```

<TableWithNestedRows>

| Parameter               | Type     | Description                                                                                                                                                                                                                                                                                 |
| ----------------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | tagMatch                | object   | It is required.                                                                                                                                                                                                                                                                             |

Reviewer 2: Can we include something more descriptive here? Maybe "The tagMatch object and its properties that are used for matching metrics. Required."

Author: Sure, added in d1f8377778213e55b22377030d206df70bdab153

> | tagMatch.name           | string   | It is a required field and it must not be an empty string. It will replace the current metric's url and name tag values when a match is found.                                                                                                                                              |

Reviewer 2: I'm a little bit confused by this description and the following ones as well. 🤔 

What happens with the value of this property? I'm wondering if we could tweak the description for this and the following values to be something along the lines of: "The name value that replaces the current metric's URL and name tag values, if a match is found. Required, and must not be an empty string."

Author: I've made the change, WDYT? 152d0b3df5d63a1969cd863a54886a16170f9801
